source/s3_meta_request.c

```diff
@@ -0,0 +1,335 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "s3_client.h"
+
+#include "auth.h"
+#include "http.h"
+#include "io.h"
+
+#include <aws/http/request_response.h>
+#include <aws/io/stream.h>
+
+static const char *s_capsule_name_s3_meta_request = "aws_s3_meta_request";
+
+struct s3_meta_request_binding {
+    struct aws_s3_meta_request *native;
+
+    bool release_called;
+    bool shutdown_called;
+
+    /* Reference proxy to python self */
+    PyObject *self_py;
+
+    /* Shutdown callback, all resource cleaned up, reference cleared after invoke */
+    PyObject *on_shutdown;
+};
+
+static void s_destroy_if_ready(struct s3_meta_request_binding *meta_request) {
+    if (meta_request->native && (!meta_request->shutdown_called || !meta_request->release_called)) {
+        /* native meta_request successfully created, but not ready to clean up yet */
+        return;
+    }
+    /* in case native never existed and shutdown never happened */
+    Py_XDECREF(meta_request->on_shutdown);
+    aws_mem_release(aws_py_get_allocator(), meta_request);
+}
+
+static void s_s3_meta_request_release(struct s3_meta_request_binding *meta_request) {
+    AWS_FATAL_ASSERT(!meta_request->release_called);
+
+    aws_s3_meta_request_release(meta_request->native);
+
+    meta_request->release_called = true;
+
+    s_destroy_if_ready(meta_request);
+}
+
+static int s_get_py_headers(const struct aws_http_headers *headers, PyObject *header_py) {
+    /* Not take the reference of header_py, caller is the one holding the reference. */
+    size_t num_headers = aws_http_headers_count(headers);
+
+    for (size_t i = 0; i < num_headers; i++) {
+        struct aws_http_header header;
+        AWS_ZERO_STRUCT(header);
+        if (aws_http_headers_get_index(headers, i, &header)) {
+            goto error;
+        }
+        const char *name_str = (const char *)header.name.ptr;
+        size_t name_len = header.name.len;
+        const char *value_str = (const char *)header.value.ptr;
+        size_t value_len = header.value.len;
+        PyObject *tuple = Py_BuildValue("(s#s#)", name_str, name_len, value_str, value_len);
+        if (!tuple) {
+            goto error;
+        }
+        PyList_SET_ITEM(header_py, i, tuple); /* steals reference to tuple */
+    }
+    return AWS_OP_SUCCESS;
+error:
+    return AWS_OP_ERR;
+}
+
+static void s_s3_request_on_headers(
+    struct aws_s3_meta_request *meta_request,
+    const struct aws_http_headers *headers,
+    int response_status,
+    void *user_data) {
+    (void)meta_request;
+    struct s3_meta_request_binding *request_binding = user_data;
+
+    /*************** GIL ACQUIRE ***************/
+    PyGILState_STATE state;
+    if (aws_py_gilstate_ensure(&state)) {
```

Comment:
Do we have to have the GIL at this point? could we delay it until we're ready to perform the callback?